core/BTB.hpp

```diff
@@ -0,0 +1,277 @@
+#pragma once
+
+#include "sparta/simulation/Unit.hpp"
+#include "sparta/simulation/TreeNode.hpp"
+#include "sparta/simulation/ParameterSet.hpp"
+
+#include "sparta/utils/MathUtils.hpp"
+
+#include "cache/TreePLRUReplacement.hpp"
+#include "cache/BasicCacheItem.hpp"
+#include "cache/Cache.hpp"
+#include "cache/BasicCacheSet.hpp"
+
+#include "MiscUtils.hpp"
+
+
+namespace olympia
+{
+
+    /**
+     * @file   BTB.hpp
+     * @brief The Branch Target Buffer -- Cache like structure for branch instructions
+     *
+     * Based on SimpleCache, this unit defines a way associate cache like structure
+     * designed to be used as a way of steering fetch ahead.
+     *
+     * Each set covers an entire block of memory, there may be multiple branches with
+     * the same block address. We use a full tag to hit on the right line.
+     *
+     * We store the branch PC, target, type and a 2 bit saturating counter to predict
+     * direction on conditional branches.
+     */
+
+    class BTBEntry : public sparta::cache::BasicCacheItem
+    {
+    public:
+
+        enum class BranchType : std::uint16_t {
+            CONDITIONAL,
+            __FIRST = CONDITIONAL,
+            DIRECT,
+            INDIRECT,
+            RETURN,
+            __LAST
+        };
+
+        BTBEntry() :
+            valid_(false)
+        { };
+
+        BTBEntry(const BTBEntry &rhs) :
+            BasicCacheItem(rhs),
+            valid_(rhs.valid_),
+            target_(rhs.target_),
+            is_call_(rhs.is_call_),
+            branchtype_(rhs.branchtype_),
+            lhist_counter_(rhs.lhist_counter_)
+        { };
+
+        BTBEntry &operator=(const BTBEntry &rhs)
+        {
+            if (&rhs != this) {
+                BasicCacheItem::operator=(rhs);
+                valid_ = rhs.valid_;
+                target_ = rhs.target_;
+                is_call_ = rhs.is_call_;
+                branchtype_ = rhs.branchtype_;
+                lhist_counter_ = rhs.lhist_counter_;
+            }
+            return *this;
+        }
```

Comment:
You can do this:
```
```suggestion
        BTBEntry() = default;
        BTBEntry(const BTBEntry &rhs) = default;
        BTBEntry &operator=(const BTBEntry &rhs) = default;
```